sotodlib/wiregrid/calibrate.py

```diff
@@ -0,0 +1,404 @@
+
+wg_degpercount = 360/52000
+stream_ids = ['ufm_mv19', 'ufm_mv18', 'ufm_mv22', 'ufm_mv29', 'ufm_mv7', 'ufm_mv9', 'ufm_mv15'] # e.g. MF1
+
+def get_det_angle(aman, **kwargs):
+    """
+    Calibrate TODs and get det_angle_raw by using the wire grid.
+    Before using this code, your axis manager must have two field implemented by HWP, i.e., demodQ and demodU
+
+    Parameters:
+    - aman : AxisManager
+    - wg_config : yaml file for the settings to load the wire grid meta data
+        e.g. wg_config = '/so/home/hnakata/scratch/metadata/context/2401_wg_satp1.yaml'
+    - wg_offset : the offset angle of wires defined by the hardware of the Grid Loader
+    - stopped_time : the stopped time of your target operation
+    Return:
+    - aman : wraped with the calibration data of the wire grid. Fields are explained in each method.
+    """
+    _config = kwargs['wg_config']
+    _offset = kwargs['wg_offset']
+    _stopped_time = kwargs['stopped_time']
+    #
+    # wrap the house-keeping data of the wire grid
+    aman = wg_wrap_hk(aman, _config)
+    #
+    # restrict AxisManger within the operation range
+    idx_wg_inside = _get_operation_range(aman)
+    aman = aman.restrict('samps', (idx_wg_inside[0], idx_wg_inside[-1]), in_place=False)
+    #
+    # wrap the demodQ/demodU related with the steps of the wire grid
+    # AxisManager here must have demodQ/demodU by the HWP
+    aman = wg_wrap_QU(aman, _stopped_time)
+    #
+    # fit the QU signal created by the wire grid
+    aman = wg_get_cfitres(aman)
+    aman = wg_get_lfitres(aman)
+    #
+    # get the det angles
+
+    det_angle_raw = 0.5*(aman.wg.lfitval[:,1] + aman.wg.lfitval[:,0]*(_offset))%180
+    det_angle_raw_err = 0.5*aman.wg.lfiterr[:,1] # temporal value
+    #
+    aman.wg.wrap('det_angle_raw', det_angle_raw, [(0, 'dets')])
+    aman.wg.wrap('det_angle_raw_err', det_angle_raw_err, [(0, 'dets')])
+    #
+    return aman
+
+# wrap HouseKeeping data
+def wg_wrap_hk(aman, wg_config):
+    """
+    Wrap the house-keeping data about the wire grid operation.
+
+    Parameters:
+    - aman : AxisManager
+    - wg_config : same as the definition in the get_det_angle
+
+    Return:
+    - aman : AxisManager
+        This includes fields, which are related with the wire grid hardware.
+        enc_deg : Encoder counts in degree of the rotation
+        LSL1 : Status of the limit switch LEFT 1 of the actuator
+        LSL2 : Status of the limit switch LEFT 2 of the actuator
+        LSR1 : Status of the limit switch RIGHT 1 of the actuator
+        LSR2 : Status of the limit switch RIGHT 2 of the actuator
+    """
+    wgctx = Context(wg_config)
+    metakeys = wgctx['metakey']
+    wgfield = wgctx['wgfield']
+    agent_names = wgctx['wg_agent_names']
+    enc = agent_names['encoder']
+    act = agent_names['actuator']
+    #
+    wgenc = hk_utils.get_detcosamp_hkaman(aman, alias=['enc_deg'], fields = [wgfield['encoder']], data_dir = metakeys['hk_dir'])
```

Comment:
I think these 5 calls to get_detcosamp_hkaman should be doable in 1 if you pass a list to `fields` instead of a single field each time. Also the alias for lsl1, lsl2, lsr1, and lsr2 are all set to `limitswitchL1`

Reply:
I didn't know well this co-sample code. I'll check it to find a more simple description.

Reply:
yes, it should be able to accept a slide of fields. And if it can't then we should fix it so it can.